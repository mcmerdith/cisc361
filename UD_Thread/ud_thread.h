/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
