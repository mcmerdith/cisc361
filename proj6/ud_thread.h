/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;