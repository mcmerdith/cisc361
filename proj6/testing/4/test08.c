/*
 * Test Program #8 - Asynchronous Send/Receive
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ud_thread.h"

void producer(int val)
{
    char msg[] = "hello world...";
    int i = 0;

    for (i = 0; i < 4; i++)
    {
        msg[12] = (char)i + 0x30;
        send(2, msg, strlen(msg));
        printf("thread %d [%s]...\n", val, msg);
    }

    printf("producer thread %d terminates...\n", val);
    t_terminate();
}

void consumer(int val)
{
    int i, len, who = 0;
    char *msg;

    msg = (char *)malloc(1024);

    for (i = 0; i < 8; i++)
    {
        who = 0;
        receive(&who, msg, &len);
        if (who != 0)
            printf("I got message [%s] from %d...\n", msg, who);
    }

    free(msg);

    printf("consumer thread %d terminate...\n", val);
    t_terminate();
}

int main(void)
{
    t_init();

    t_create(producer, 1, 1);
    t_create(producer, 3, 1);
    t_create(consumer, 2, 1);

    t_yield();

    printf("thread library shutting down...\n");
    t_shutdown();

    return 0;
}