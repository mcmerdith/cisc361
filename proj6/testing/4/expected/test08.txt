thread 1 [hello world.0.]...
thread 1 [hello world.1.]...
thread 1 [hello world.2.]...
thread 1 [hello world.3.]...
producer thread 1 terminates...
thread 3 [hello world.0.]...
thread 3 [hello world.1.]...
thread 3 [hello world.2.]...
thread 3 [hello world.3.]...
producer thread 3 terminates...
I got message [hello world.0.] from 1...
I got message [hello world.1.] from 1...
I got message [hello world.2.] from 1...
I got message [hello world.3.] from 1...
I got message [hello world.0.] from 3...
I got message [hello world.1.] from 3...
I got message [hello world.2.] from 3...
I got message [hello world.3.] from 3...
consumer thread 2 terminate...
thread library shutting down...
